### PR TITLE
#3129 - Full template preview following mouse cursor

### PR DESCRIPTION
**Pull request by KonstantinEpam23**
## How the feature works? / How did you fix the issue?
https://github.com/epam/ketcher/assets/16281184/14d0b744-43dd-4a33-9462-4b4dab9dcce2

## Check list
- [ ] unit-tests written
- [x] e2e-tests written
- [x] documentation updated
- [x] PR name follows the pattern `#1234 – issue name`
- [x] branch name doesn't contain '#'
- [x] PR is linked with the issue
- [x] base branch (master or release/xx) is correct
- [x] task status changed to "Code review"
- [x] reviewers are notified about the pull request

### CODE REVIEW

**Comment by KonstantinEpam23** [comment]
The current issue is that after selecting the template - if the user presses undo - the template preview will be left on the canvas instead of hiding the preview and undoing the last action.

I have started re-doing the approach in branch: `3129-full-template-preview-following-mouse-cursor-draft-4`

File: packages/ketcher-react/src/script/editor/tool/templatePreview.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  fromTemplateOnCanvas,
+  fromMultipleMove,
+  getHoverToFuse,
+} from 'ketcher-core';
+import Editor from '../Editor';
+import { MODES } from 'src/constants';
+import { getAngleFromEvent, getBondFlipSign } from './template';
+
+const PREVIEW_DELAY = 300;
+type CiType = { map: string; id: number; dist: number };
```
Comment (nanoblit) [inline packages/ketcher-react/src/script/editor/tool/templatePreview.ts:33]: What's "ci"? Maybe we can use the full name?

**Comment by Zhirnoff** [comment]
16 tests fail. Please add screenshots of local run

**Review by rrodionov91 — CHANGES_REQUESTED** [review]
Please check this behaviour. Looks like bugs.

https://github.com/epam/ketcher/assets/14859362/d2fa9555-16a7-46c2-a8c9-744e05b8711a



https://github.com/epam/ketcher/assets/14859362/be7c7b50-cf3d-4150-8647-96755893d332


https://github.com/epam/ketcher/assets/14859362/d696d0b3-b572-4559-9b09-feb1f518b4f9

**Comment by Zhirnoff** [comment]
Pre-tested. Bug is fixed. And fixed comments added by @rrodionov91

File: ketcher-autotests/tests/User-Interface/Preview-For-Abbreviated-Structures/preview-for-abbreviated-structures.spec.ts [hunk 2]
```diff
@@ -96,4 +96,41 @@ test.describe('Preview for abbreviated structures: functional groups', () => {
...
+      { type: BondType.SINGLE },
+      bondId,
+    );
+    const pointAwayFromBond = {
+      x: bondPosition.x + shift,
+      y: bondPosition.y + shift,
+    };
+    await takeEditorScreenshot(page);
+    await page.mouse.move(pointAwayFromBond.x, pointAwayFromBond.y);
+    await takeEditorScreenshot(page);
```
Comment (rrodionov91) [inline ketcher-autotests/tests/User-Interface/Preview-For-Abbreviated-Structures/preview-for-abbreviated-structures.spec.ts:134]: I believe that this test covers the case from the previous one and we can keep just only this test.

File: packages/ketcher-react/src/script/editor/tool/templatePreview.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+type ClosestItemType = { map: string; id: number; dist: number };
+
+function getUniqueCiId(ci: ClosestItemType) {
+  return `${ci.id}-${ci.map}`;
+}
+
+class TemplatePreview {
+  private readonly editor: Editor;
+  private readonly template;
+  private readonly mode: 'fg' | null;
```
Comment (rrodionov91) [inline packages/ketcher-react/src/script/editor/tool/templatePreview.ts:42]: I believe we can have more clear values for "mode".

File: packages/ketcher-react/src/script/editor/tool/templatePreview.ts [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+      return ci;
+    } else {
+      return null;
+    }
+  }
+
+  movePreview(event: PointerEvent) {
+    this.position = this.editor.render.page2obj(event);
+
+    const ci = this.getPreviewTarget();
```
Comment (rrodionov91) [inline packages/ketcher-react/src/script/editor/tool/templatePreview.ts:102]: I would propose to rename ci to previewTarget

File: packages/ketcher-react/src/script/editor/tool/templatePreview.ts [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    const isMouseAwayFromAtomsAndBonds = !ci;
+    const isPreviewTargetChanged =
+      ci && this.lastPreviewId !== getUniqueCiId(ci);
+
+    const shouldHidePreview =
+      isMouseAwayFromAtomsAndBonds || isPreviewTargetChanged;
+
+    const shouldShowPreview =
+      ci && !this.connectedPreviewAction && !this.connectedPreviewTimeout;
```
Comment (rrodionov91) [inline packages/ketcher-react/src/script/editor/tool/templatePreview.ts:111]: I would propose to rename it to shouldHideConnectedPreview and shouldShowConnectedPreview

File: packages/ketcher-react/src/script/editor/tool/templatePreview.ts [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        this.showFloatingPreview(this.position);
+        this.previousPosition = this.position;
+        this.editor.render.update(false, null, { resizeCanvas: false });
+      } else {
+        const dist = this.position.sub(this.previousPosition);
+        this.previousPosition = this.position;
+        fromMultipleMove(this.restruct, this.floatingPreview, dist);
+        this.editor.render.update(false, null, { resizeCanvas: false });
+        this.hoverFusedItems(ci, event);
+      }
```
Comment (rrodionov91) [inline packages/ketcher-react/src/script/editor/tool/templatePreview.ts:135]: I would propose to move it to separate function f.e. moveFloatingPreview()

**Review by rrodionov91 — APPROVED** [review]

**Pull request merged**